src/server/images/controllers.js

```diff
@@ -134,5 +134,123 @@ export default {
     .catch((err) => {
       return cb(new APIError(err, {statusCode: err.statusCode, message: err.message}));
     });
+  },
+  remove(req, res, cb) {
+    const {userKey} = req.headers;
+    const imageKey = req.params.id;
+    assert(userKey, 'user-key should be provided');
+    assert(imageKey, 'image-key should be provided');
+
+    let itemKey;
+    let imageIndexKey;
+    let createdPostKey;
+    let imageIndexValue;
+    let createdPostValue;
+
+    getPromise(imageKey).then((value) => {
+      // make keys
+      itemKey = value.itemKey;
+      const timeHash = imageKey.slice(6);
+      imageIndexKey = KeyUtils.getIdxKey(ENTITY.IMAGE, timeHash, itemKey);
+      createdPostKey = KeyUtils.getIdxKey(ENTITY.CREATED_POST, timeHash, userKey);
+    }).then(() => {
+      // save values
+      return getPromise(imageIndexKey).then((value) => {
+        imageIndexValue = value;
+        return getPromise(createdPostKey);
+      }).then((value) => {
+        createdPostValue = value;
+      });
+    }).then(() => {
+      // delete ALIVE state
+      const opts = [
+        { type: 'del', key: imageIndexKey },
+        { type: 'del', key: createdPostKey }
+      ];
+      return new Promise((resolve, reject) => {
+        db.batch(opts, (err) => {
+          return err ? reject(err) : resolve();
+        });
+      });
+    }).then(() => {
+      // add REMOVED state
+      function replaceAt(str, index, char) {
+        // TODO: Refactoring : This function is duplicated with replaceState
+        return str.substr(0, index) + char + str.substr(index + char.length);
+      }
+      const removedImageIndexKey =
+        replaceAt(imageIndexKey, imageIndexKey.indexOf('0'), STATE.REMOVED);
+      const removedCreatedPostKey =
+        replaceAt(createdPostKey, createdPostKey.indexOf('0'), STATE.REMOVED);
+      const opts = [
+        { type: 'put', key: removedImageIndexKey, value: imageIndexValue },
+        { type: 'put', key: removedCreatedPostKey, value: createdPostValue }
+      ];
+      return new Promise((resolve, reject) => {
+        db.batch(opts, (err) => {
+          return err ? reject(err) : resolve();
+        });
+      });
+    }).then(() => {
+      // count image of item
+      const prefixes = [];
+      const promises = [];
+      prefixes.push(KeyUtils.getPrefix(ENTITY.IMAGE, STATE.ALIVE, itemKey));
+      prefixes.push(KeyUtils.getPrefix(ENTITY.IMAGE, STATE.EXPIRED, itemKey));
+
+      prefixes.map((prefix) => {
```

Comment:
you can just return promise.all(prefixes.map((prefix) => { return new Promise() }));